[iOS] 로그인 기능을 구현

ehgud0670 opened the pull request:
### 로그인 기능을 구현하였습니다.
> Issue #103  , #104 , #113 을 구현하였습니다.

* 로그인 화면을 구현하였습니다.
* Github OAuth 인증 기능을 구현하였습니다.
* 조건에 따라 로그인 화면을 띄우도록 구현하였습니다.
  * 토큰을 가지고 있지 않으면 로그인 화면을 띄우고 로그인을 하면 델리게이트 패턴을 통해 검색화면에서 모델 데이터를 fetch 하도록 했습니다.
  * 토큰을 가지고 있으면 검색화면에서 로그인 화면을 띄우지 않고, 모델 데이터를 fetch 하도록 했습니다.

> 실행 화면

![login](https://user-images.githubusercontent.com/38216027/83595180-31f76700-a59c-11ea-987f-fd7df9c56bed.gif)

ehgud0670 committed:
[#103] feat: 로그인 화면에 대한 VC, 스토리보드 구현하고 SearchVC에서 띄우도록 함

ehgud0670 committed:
[#103] feat: imageView를 추가하고 airbnb 타이틀 + 로고가 있는 이미지 적용

ehgud0670 committed:
[#103] feat: CornerRoundedButton을 상속받은 로그인 버튼을 구현하여 화면에 추가
 	     highlighted 경우의 titleColor를 설정해야 하기 때문에 어쩔 수 없이
	     CornerRoundedButton을 상속하고 highlighted 경우의 titleColor를 설정했다.

ehgud0670 committed:
[#103] feat: 클로즈 버튼을 화면에 추가하고 누르면 화면이 dismiss되도록 기능 구현함

ehgud0670 committed:
[#103] refactor: 로그인 화면 present 하는 코드 삭제
	이유: 로그인 기능을 모두 구현하지 않아서 아직 화면을 보일 필요는 없다.

ehgud0670 committed:
[#104] feat: LoginButton 액션 추가 및 Web Authentication Session 생성

ehgud0670 committed:
[#104] feat: Persentation Context 제공

ehgud0670 committed:
[#104] feat: callBack 함수 처리해서 jwt 토큰 빼오기

ehgud0670 committed:
[#??] feat: ASWebAuthenticationSession객체가 동작하도록 start 메소드 호출

ehgud0670 committed:
[#104] refactor: 역할이 분리되어 있는 것 같아 메소드를 합침

ehgud0670 committed:
[#104] feat:  clientID, scheme, payloadKey 관련 파일 추가

ehgud0670 committed:
[#104] docs: AuthKeys.swift path 삭제

ehgud0670 committed:
[#104] feat: callbackURL이 오면 거기서 jwt 토큰을 추출해 UserDefault에 저장하고
	     화면이 dismiss 되도록 기능 구현함

ehgud0670 committed:
[#113] feat: jwt 토큰이 없는 경우에 로그인 화면을 띄우도록 기능 구현
	* 가지고 있는 토큰이 검증되있지 않은 경우는 다음 커밋에서 할 예정이다.

ehgud0670 committed:
[#113] refactor: 해당 jwt 토큰이 유효한지 검사하는 로직 삭제
	이유: 백엔드와의 토의 결과 jwt 토큰은 무기한이기 때문에 일단은 유효한지 검증하는 로직은 삭제하기 했다.

ehgud0670 committed:
[#113] refactor: 토큰이 있으면 데이터를 fetch하고,
		 토큰이 없으면 로그인 화면을 띄우고 로그인이 된 경우에 검색 화면에서 데이터를 fetch 하도록 기능 수정함

ehgud0670 committed:
[#103] chore: 줄바꿈 옳게 되도록 수정

seizze committed:
Merge pull request #105 from codesquad-member-2020/iOS/feature/issue-103

[iOS] 로그인 화면을 구성하였습니다.

ehgud0670 committed:
[#104] style: 철차 수정

ehgud0670 committed:
[#104] refactor: AuthRequest를 만들어 재적용

seizze committed:
[#104] refactor: AuthRequest 변경

seizze committed:
Merge pull request #112 from codesquad-member-2020/iOS/feature/issue-104

[iOS] github OAuth 인증 후 jwt 토큰을 UseDefaults에 저장하는 기능을 구현하였습니다.

ehgud0670 committed:
[#113] refactor: 로그인 뷰 컨트롤러를 파라미터로 추가
	   이유: 델리게이트 메시지가 어디서 오는지 알기 위해 추가하였다.

seizze committed:
Merge pull request #127 from codesquad-member-2020/iOS/feature/issue-113

[iOS] 조건에 따라 로그인 화면을 띄우는 기능 구현

ehgud0670 committed:
chore: resolve conflicts
Merge branch 'dev' into iOS/Login

godrm commented on iOS/Airbnb/Airbnb/Controllers/SearchViewController.swift:
```diff
@@ -31,10 +32,20 @@ final class SearchViewController: UIViewController {
+            let loginViewController = LoginViewController.instantiate(from: .login) {
+            loginViewController.delegate = self
+            present(loginViewController, animated: false)
+            hasBeenDisplayed = true
```
이렇게 SearchViewController에서 Login을 띄우는 방법도 있지만,
App 시작할 때 AppDelegate에서 Token을 확인하고 Login 화면을 띄운 다음에 로그인 하면 SearchViewController를 띄우는 방식은 어떨까요?
그러면 SearchVC가 좀 더 독립적으로 동작하지 않을까요?

ehgud0670 replied:
오 완전 좋은 방식입니다. 반영을 해보도록 하겠습니다 😀

ehgud0670 replied:
네 반영하였습니다. 감사합니다

godrm commented on iOS/Airbnb/Airbnb/Controllers/SearchViewController.swift:
```diff
@@ -87,4 +98,14 @@ final class SearchViewController: UIViewController {
+
+extension SearchViewController: LoginViewControllerDelegate {
+    func loginDidSuccess(_ viewController: LoginViewController) {
+        fetchBNBs()
```
fetch 하는 시점도 로그인 이후에 할건지 로그인 이전에 할 건지 고민해볼 필요가 있습니다.
애플도 앱이 로그인 하기 전에 (심지어 가입하기도 전에) 앱을 둘러볼 수 있도록 만들기를 권장합니다.
결제가 필요한 시점에부터만 로그인을 하도록 유도하는거죠. 
그럴려면 화면 ViewController가 어느 시점에서든지 동작할 수 있도록 하는 것도 좋습니다.

ehgud0670 replied:
오 넵 피드백을 고려하여 viewDidLoad에서 모델 데이터를 fetch 하도록 하였습니다. 감사합니다.

ehgud0670 committed:
[#login] refactor: 1. sceneDelegate에서 jwt이 있으면 검색화면을 띄우고,
		   없으면 로그인화면을 띄우도록 함
		   2. 로그인이 되든, 안되든 모델 데이터를 fetch 하도록 ViewDidLoad에서 모델을 fetch하도록 코드 변경함
